File: 3.5 Dynamic Programming/Coin Change/p11517.py
# Competitive Programming 3
# Problem 11517

from math import inf

T = int(input())

for _ in range(T):
    V = int(input())
    
    n = int(input())
    notes = [0 for _ in range(n)]

    for i in range(n):
        notes[i] = int(input())
        
    notes.sort()
    
    # number of bills required to make value x
    dp = [inf for _ in range(sum(notes) + 1)]
    dp[0] = 0
    
    for j in range(n):
        note = notes[j]
        for i in range(sum(notes[:j]), -1, -1):
            if dp[i] != inf:
                dp[i+note] = min(dp[i+note], dp[i] + 1)
        
    i = V
    while True:     
        if dp[i] != inf:
            print(i, dp[i])
            break
        i += 1        

File: 3.5 Dynamic Programming/Coin Change/p11517.cpp
// Competitive Programming 3
// Problem 11517

#include <iostream>
#include <cstring>
using namespace std;

#define INF (1<<30)

int main(){
    int T;
    cin >> T;

    while (T--){
        int V, n, i;
        cin >> V;
        cin >> n;

        int notes[n];
        for (i = 0; i < n; i++)
            cin >> notes[i];

        int total = V + 10001;
        int dp[total];
        for (int i = 0; i < total; i++) dp[i] = INF;
        dp[0] = 0;

        for (int n : notes){
            for (int i = total - n + 1; i >= 0; i--){
                if (dp[i] != INF)
                    dp[i+n] = min(dp[i+n], dp[i] + 1);
            }
        }

        i = V;
        while (dp[i] == INF)
            i++;
        printf("%d %d\n", i, dp[i]);
    }
    return 0;
}
